flowauth/backend/tests/test_token_generation.py

```diff
@@ -9,77 +9,114 @@
-    uid, uname, upass = test_user
-    # Log in first
-    response, csrf_cookie = auth.login(uname, upass)
-    expiry = (datetime.datetime.now() + datetime.timedelta(minutes=2)).strftime(
-        "%Y-%m-%dT%H:%M:%S.%fZ"
-    )
-    token_eq = {"name": "TEST_TOKEN", "expiry": expiry, "claims": ["run&DUMMY_ROUTE_A"]}
-    response = client.post(
-        "/tokens/tokens/1", headers={"X-CSRF-Token": csrf_cookie}, json=token_eq
-    )
-    assert 401 == response.status_code
-    assert (
-        b"You do not have access to run&DUMMY_ROUTE_A on DUMMY_SERVER_A"
-        in response.get_data()
-    )
 
+@freeze_time(datetime.datetime(year=2020, month=12, day=31))
+def test_reject_when_claim_not_allowed(
+    client, auth, app, test_user, test_roles, test_scopes
+):
+    with app.app_context():
+        uid, uname, upass = test_user
+        runner_role, reader_role, reader_b_role = test_roles
 
-@pytest.mark.usefixtures("test_data_with_access_rights")
-def test_token_generation(client, auth, app, test_user, public_key):
-    # Log in first
-    uid, uname, upass = test_user
-    response, csrf_cookie = auth.login(uname, upass)
-    expiry = datetime.datetime.now() + datetime.timedelta(minutes=2)
-    token_eq = {
-        "name": "DUMMY_TOKEN",
-        "expiry": expiry.strftime("%Y-%m-%dT%H:%M:%S.%fZ"),
-        "claims": [
-            "run&DUMMY_ROUTE_A.aggregation_unit.admin0",
-            "get_result&DUMMY_ROUTE_A.aggregation_unit.admin0",
-        ],
-    }
-    response = client.post(
-        "/tokens/tokens/1", headers={"X-CSRF-Token": csrf_cookie}, json=token_eq
-    )
-    assert 200 == response.status_code
-    token_json = response.get_json()
-    decoded_token = jwt.decode(
-        jwt=token_json["token"].encode(),
-        key=public_key,
-        algorithms=["RS256"],
-        audience="DUMMY_SERVER_A",
-    )
-    assert [
-        "get_result,run&DUMMY_ROUTE_A.aggregation_unit.admin0"
-    ] == decompress_claims(decoded_token["user_claims"])
-    assert "TEST_USER" == decoded_token["identity"]
-    assert approx(expiry.timestamp()) == decoded_token["exp"]
+        # Give test_user some roles on DUMMY_SERVER_A
+        test_user = db.session.execute(db.select(User).where(User.id == uid)).scalar()
+        test_user.roles.append(reader_role)
+
+        # Log in first
+        response, csrf_cookie = auth.login(uname, upass)
+        token_eq = {
+            "name": "TEST_TOKEN",
+            "roles": [{"name": "runner"}],
+        }
+        response = client.post(
+            "/tokens/tokens/1", headers={"X-CSRF-Token": csrf_cookie}, json=token_eq
+        )
+        assert 401 == response.status_code
+        assert (
+            "Role 'runner' is not permitted for the current user"
+            == response.get_json()["message"]
+        )
+
+        # Testing attempting reader on second server
+        response = client.post(
+            "/tokens/tokens/2", headers={"X-CSRF-Token": csrf_cookie}, json=token_eq
```

Comment:
I think you want to request a token for `reader_role` here (i.e. a role that test_user does have access to, but for the wrong server). Currently `token_eq` is unchanged from the first part of the test.